core/services/offchainreporting/data_source.go

```diff
@@ -22,16 +24,43 @@ type dataSource struct {
 
 var _ ocrtypes.DataSource = (*dataSource)(nil)
 
-// The context passed in here has a timeout of observationTimeout.
-// Gorm/pgx doesn't return a helpful error upon cancellation, so we manually check for cancellation and return a
-// appropriate error (FIXME: How does this work after the current refactoring?)
+// The context passed in here has a timeout of (ObservationTimeout - ObservationGracePeriod).
+// Upon context cancellation, its expected that we return any usable values within ObservationGracePeriod.
 func (ds dataSource) Observe(ctx context.Context) (ocrtypes.Observation, error) {
-	results, err := ds.pipelineRunner.ExecuteAndInsertNewRun(ctx, ds.spec, ds.ocrLogger)
+	var observation ocrtypes.Observation
+	start := time.Now()
+	run, err := pipeline.NewRun(ds.spec, start)
 	if err != nil {
-		return nil, errors.Wrapf(err, "error executing new run for job ID %v", ds.jobID)
+		return observation, errors.Wrapf(err, "error creating new run for spec ID %v", ds.spec.ID)
 	}
 
-	result, err := results.SingularResult()
+	trrs, err := ds.pipelineRunner.ExecuteRun(ctx, run, ds.ocrLogger)
+	if err != nil {
+		return observation, errors.Wrapf(err, "error executing run for spec ID %v", ds.spec.ID)
+	}
+	end := time.Now()
+
+	run.FinishedAt = &end
+
+	finalResult := trrs.FinalResult()
+	run.Outputs = finalResult.OutputsDB()
+	run.Errors = finalResult.ErrorsDB()
+
+	// Do the database write in a non-blocking fashion
+	// so we can return the observation results immediately.
+	// This is helpful in the case of a blocking API call, where
+	// we reach the passed in context deadline and we want to
+	// immediately return any result we have and do not want to have
+	// a db write block that.
+	go func() {
+		// In the case that our run gets cancelled via ctx cancellation,
+		// we still want to save the results and so we need a fresh context.
+		if _, err := ds.pipelineRunner.InsertFinishedRunWithResults(context.Background(), run, trrs); err != nil {
```

Comment:
I would be happier to see _some_ sanity timeout on this.

Reply:
The timeout is here https://github.com/smartcontractkit/chainlink/blob/72b8d6e6525c7a6469686eeac88b8f41e209a5d0/core/services/pipeline/runner.go#L462